feat(db_api): support stale reads

IlyaFaer opened the pull request:
Based on #475 and requires it to be merged first.

Commit:
feat(db_api): add an ability to set ReadOnly/ReadWrite connection mode

Commit:
update the error message

Commit:
update another error message

Commit:
don't check error messages with regexes

Commit:
don't commit/rollback ReadOnly transactions

Commit:
clear the transaction

Commit:
add read-only transactions data visibility test

Commit:
Apply suggestions from code review

Co-authored-by: Knut Olav Løite <koloite@gmail.com>

Commit:
add conditions for edge cases

Commit:
Merge branch 'read_only_transactions' of https://github.com/q-logic/python-spanner into read_only_transactions

Commit:
don't calc checksum for read-only transactions

Commit:
Merge branch 'master' into read_only_transactions

Commit:
use Snapshot for reads

Commit:
update docstrings

Commit:
Merge branch 'master' into read_only_transactions

Commit:
Merge branch 'master' into read_only_transactions

Commit:
use multi-use snapshots in !autocommit mode

Commit:
return the read_only docstring back, erase excess unit test

larkee committed:
Merge branch 'master' into read_only_transactions

Commit:
erase excess ifs

Commit:
Merge branch 'read_only_transactions' of https://github.com/q-logic/python-spanner into read_only_transactions

Commit:
Merge branch 'master' into read_only_transactions

Commit:
add additional check into the snapshot_checkout() method

Commit:
Merge branch 'master' into read_only_transactions

Commit:
add new style system test

Commit:
Merge branch 'read_only_transactions' of https://github.com/q-logic/python-spanner into read_only_transactions

Commit:
resolve conflict

Commit:
don't use error message regexes

Commit:
erase excess import

Commit:
Merge branch 'master' into read_only_transactions

Commit:
refactor

larkee committed:
Merge branch 'main' into read_only_transactions

Commit:
add unit test to check that read-only transactions are not retried

Commit:
Merge branch 'read_only_transactions' of https://github.com/q-logic/python-spanner into read_only_transactions

Commit:
Merge branch 'read_only_transactions' into stale_reads

Commit:
feat(db_api): support stale reads

Commit:
Merge branch 'master' into stale_reads

larkee commented on google/cloud/spanner_dbapi/connection.py:
```diff
@@ -165,6 +166,48 @@ def read_only(self, value):
+        Returns:
+            dict: Staleness type and value.
+        """
+        return self._staleness
```
In the interest of avoiding having to add `or {}` to a bunch of checks, perhaps it makes more sense to do one of the following:

1) Keep the default for `self._staleness` as `None` and update this line to `return self._staleness or {}`
2) Change the default for `self._staleness` to `{}` and update this method to return `None` if `self._staleness is {}`

Which one we use depends on whether getting `None` or an empty dict is more useful/clearer when calling `conn.staleness`

larkee commented on tests/unit/spanner_dbapi/test_connection.py:
```diff
@@ -763,3 +755,121 @@ def test_validate_closed(self):
+
+    def test_staleness_inside_transaction(self):
+        """
+        Check that it's impossible to change the `steleness`
```
```suggestion
        Check that it's impossible to change the `staleness`
```

larkee commented on tests/unit/spanner_dbapi/test_connection.py:
```diff
@@ -763,3 +755,121 @@ def test_validate_closed(self):
+
+    def test_staleness_multi_use(self):
+        """
+        Check that `steleness` option is correctly
```
```suggestion
        Check that `staleness` option is correctly
```

Commit:
Apply suggestions from code review

Co-authored-by: larkee <31196561+larkee@users.noreply.github.com>

Commit:
move `or None` into the property

Commit:
Merge branch 'stale_reads' of https://github.com/q-logic/python-spanner into stale_reads

Commit:
Merge branch 'main' into stale_reads

Commit:
Merge branch 'main' into stale_reads

larkee committed:
Merge branch 'main' into stale_reads

Commit:
fix the check

IlyaFaer commented on google/cloud/spanner_dbapi/connection.py:
```diff
@@ -165,6 +166,48 @@ def read_only(self, value):
+                    one_of_expected = True
+                    break
+
+            if not one_of_expected:
```
Done

Commit:
lint fix

skuruppu commented:
@IlyaFaer what do we need to do to get this merged? Are the test failures related to this change or are they flaky?

Commit:
lint fix

IlyaFaer commented:
@skuruppu, looks like there was a problem with lint (I've updated the `black` linting package locally, but kokoro seems to be using a particular version). Fixed it, let's see if everything is okay now.

larkee committed:
Merge branch 'main' into stale_reads

larkee committed:
Merge branch 'main' into stale_reads

Commit:
Merge branch 'main' into stale_reads

larkee committed:
Merge branch 'main' into stale_reads
